internal/controllers/machine/machine_controller_noderef.go

```diff
@@ -131,6 +144,50 @@
 	return ctrl.Result{}, nil
 }
 
+// unstructuredNode returns a raw unstructured from a Node.
+func unstructuredNode(node *corev1.Node) *unstructured.Unstructured {
+	obj := &unstructured.Unstructured{}
+	obj.SetAPIVersion(node.GroupVersionKind().GroupVersion().String())
+	obj.SetKind(node.Kind)
+	obj.SetName(node.Name)
+	obj.SetUID(node.UID)
+	obj.SetLabels(node.Labels)
+	return obj
+}
+
+// getManagedLabels gets a map[string]string and returns another map[string]string
+// filtering out labels not managed by CAPI.
+func getManagedLabels(labels map[string]string) map[string]string {
+	managedLabels := make(map[string]string)
+	for key, value := range labels {
+		if strings.Contains(key, clusterv1.NodeRestrictionLabelDomain+"/") || strings.HasSuffix(key, clusterv1.NodeRestrictionLabelDomain) {
+			// Skip if it's not a valid subdomain
+			if !strings.HasPrefix(key, clusterv1.NodeRestrictionLabelDomain) {
+				if !strings.Contains(key, "."+clusterv1.NodeRestrictionLabelDomain) {
+					continue
+				}
+			}
+			managedLabels[key] = value
+		}
+
+		if strings.Contains(key, clusterv1.ManagedNodeLabelDomain+"/") || strings.HasSuffix(key, clusterv1.ManagedNodeLabelDomain) {
+			// Skip if it's not a valid subdomain
+			if !strings.HasPrefix(key, clusterv1.ManagedNodeLabelDomain) {
+				if !strings.Contains(key, "."+clusterv1.ManagedNodeLabelDomain) {
+					continue
+				}
+			}
+			managedLabels[key] = value
+		}
+
+		if strings.HasPrefix(key, clusterv1.NodeRoleLabelPrefix+"/") || strings.EqualFold(key, clusterv1.NodeRoleLabelPrefix) {
+			managedLabels[key] = value
+		}
+	}
```

Comment:
I found another edge case that could cause a problem. The label `custom.company.xyz/test.node-restriction.kubernetes.io` should not be among the managed labels but it will.

How about doing something like this?
```suggestion
   for key, value := range labels {
		dnsSubdomainOrName := strings.Split(key, "/")[0]
		if dnsSubdomainOrName == clusterv1.NodeRoleLabelPrefix {
			managedLabels[key] = value
		}
		if dnsSubdomainOrName == clusterv1.NodeRestrictionLabelDomain || strings.HasSuffix(dnsSubdomainOrName, "."+clusterv1.NodeRestrictionLabelDomain) {
			managedLabels[key] = value
		}
		if dnsSubdomainOrName == clusterv1.ManagedNodeLabelDomain || strings.HasSuffix(dnsSubdomainOrName, "."+clusterv1.ManagedNodeLabelDomain) {
			managedLabels[key] = value
		}
	}
```

Reply:
Thank you!